src/box/txn_limbo.c

```diff
@@ -1086,6 +1091,84 @@ txn_limbo_filter_request(struct txn_limbo *limbo,
 	}
 }
 
+static uint32_t sync_system_space_ids[] = {
+	/* BOX_VINYL_DEFERRED_DELETE_ID, local space*/
+	BOX_SCHEMA_ID,
+	BOX_COLLATION_ID,
+	BOX_VCOLLATION_ID,
+	BOX_SPACE_ID,
+	BOX_VSPACE_ID,
+	BOX_SEQUENCE_ID,
+	/* BOX_SEQUENCE_DATA_ID, synchronized by synchronous space operations */
+	BOX_VSEQUENCE_ID,
+	BOX_INDEX_ID,
+	BOX_VINDEX_ID,
+	BOX_FUNC_ID,
+	BOX_VFUNC_ID,
+	BOX_USER_ID,
+	BOX_VUSER_ID,
+	BOX_PRIV_ID,
+	BOX_VPRIV_ID,
+	BOX_CLUSTER_ID,
+	BOX_TRIGGER_ID,
+	BOX_TRUNCATE_ID,
+	BOX_SPACE_SEQUENCE_ID,
+	BOX_VSPACE_SEQUENCE_ID,
+	BOX_FK_CONSTRAINT_ID,
+	BOX_CK_CONSTRAINT_ID,
+	BOX_FUNC_INDEX_ID,
+	BOX_SESSION_SETTINGS_ID,
+};
+
+/**
+ * Controls whether to consider system spaces indefinitely synchronous when the
+ * synchronous queue is claimed.
+ */
+bool box_consider_system_spaces_synchronous;
+TWEAK_BOOL(box_consider_system_spaces_synchronous);
+
+/**
+ * Update the state of synchronous replication for system spaces in the
+ * `sync_system_space_ids` list.
+ */
+static void
+txn_limbo_update_sync_system_spaces(struct txn_limbo *limbo, uint16_t req_type,
+				    bool is_rollback)
+{
+	assert(req_type == IPROTO_RAFT_PROMOTE ||
+	       req_type == IPROTO_RAFT_DEMOTE);
+	if (is_rollback)
+		req_type = req_type == IPROTO_RAFT_PROMOTE ?
+			   IPROTO_RAFT_DEMOTE : IPROTO_RAFT_PROMOTE;
+	if (recovery_state <= INITIAL_RECOVERY ||
+	    (!is_rollback &&
+	     ((req_type == IPROTO_RAFT_PROMOTE &&
+	       limbo->owner_id != REPLICA_ID_NIL) ||
+	      (req_type == IPROTO_RAFT_DEMOTE &&
+	       limbo->owner_id == REPLICA_ID_NIL))))
```

Comment:
> This code block is very difficult to parse, probably we need comments here and I'd refactor it, so that every OR condition is just a separate if -> return.

Fixed.

> Why don't we apply space transitions only when we recover from the snap file. They're needed during xlog recovery?

Actually, AFAIC, it's pointless to apply space transition during recovery, so let's just leave the condition as `recovery_state != FINISHED_RECOVERY`.

> AFAICS, the second check is PROMOTE from some instance to smth else, as promote cannot be done with zero `origin_id`. Nothing to do. Last one is DEMOTE from nil to nil, also nothing to do. Am I right?

Right.